Simplenote/src/main/java/com/automattic/simplenote/NoteEditorFragment.java

```diff
@@ -1125,22 +1160,13 @@ private int newCursorLocation(String newText, String oldText, int cursorLocation
 
     @Override
     public void onTagAdded(String tag) {
-        if (mNote == null || !isAdded()) {
+        // When a tag is added, it triggers an event that sends an empty tag. For those cases or if the note is null,
+        // we avoid updating the UI and the note
+        if (mNote == null || !isAdded() || TextUtils.isEmpty(tag)) {
             return;
         }
 
-        if (mNote.getTagString() != null && tag.length() > mNote.getTagString().length()) {
-            AnalyticsTracker.track(
-                EDITOR_TAG_ADDED,
-                CATEGORY_NOTE,
-                "tag_added_to_note"
-            );
-        }
-
-        mNote.setTagString(mNote.getTagString() + String.valueOf(SPACE) + tag);
```

Comment:
In `NoteEditorViewModel` as part of `addTag()`.

Reply:
`addTag` updates `UiState` which updates the list of chips. that is why `updateTagList();` is not called.

Reply:
❓ What I meant, is the `mNote.setTagString(...)` logic where it gets to be concatenated with the `SPACE` etc. 🤔 What am I missing here, I am probably too stupid to understand it completely, so I do apologise for being persistent on that? 🙏

Reply:
Howdy 👋  @ParaskP7. No problem at all, I may have introduced a bug, so it is important to discuss it. 

> What I meant, is the `mNote.setTagString(...)` logic where it gets to be concatenated with the SPACE etc. 

The implementation of tag management in `NoteEditorFragment` handle 4 main events and used two sources of truth for tags: 1) the note object `mNote` through `Note.getTagString()`, and 2) the tags in the UI through the method `NotEditorFragment.getNoteTagsString()` (number of `Chip` views on the `ChipGroup`). The 4 main events are the following: 

1. `LOAD`: when the fragment is loaded, the tags are obtained from the note. Use `Note.getTagString()` as source of truth
2. `NOTE UPDATE`: when the note is updated from another device for example. Used `Note.getTagString()` as source of truth
3. `ADD TAG`: when a tag is added (UI event) through the callback `onTagAdded()`. Used `Note.getTagString()` as source of truth and `mNote.setTagString(mNote.getTagString() + String.valueOf(SPACE) + tag)` to update the state
4. `REMOVE TAG`: when a tag is removed (UI event). This event used `NotEditorFragment.getNoteTagsString()` as source of truth. 
5. `TAG FIELD LOSE FOCUS`: when the tag field loses focus. If the field had a valid tag, the tag was added and used `Note.getTagString()` as source of truth. Otherwise, it used the `NotEditorFragment.getNoteTagsString()` as source of truth to set the tags (which was redundant since the tags did not change in this case).

The whole add space and split by space was needed because the method `NotEditorFragment.getNoteTagsString()` returned a `String` of tags separated by space, instead of returning a list of tags. This drove the whole architecture to use `String` as list of tags and add or split/join by `space` in several parts of the code. 

With the refactor we drive the UI from the lower layers (`ViewModel` -> `UseCase` -> `Repository` -> `Note`). We have only one source of truth which is the `Note` modified by the `UseCase` to filter collaborators out. 

We also reduced the complexity because the fragment just deals with calling the `ViewModel` in each of the 4 events previously mentioned and reacting to change on the `UiState` to update the tags (`Chip` views). Now the 4 events are handled as follows:

1. `LOAD`: call `viewModel.update(mNote)` which updates the `UiState` that is handled by the fragment to update the tags (`Chip` views)
2. `NOTE UPDATE`: call `viewModel.update(mNote)` which updates the `UiState` that is handled by the fragment to update the tags (`Chip` views)
3. `ADD TAG`: call `viewModel.addTag(tag, mNote)` which updates the tags in the note and the `UiState` that is handled by the fragment to update the tags (`Chip` views)
4. `REMOVE TAG`: call `viewModel.removeTag(tag, mNote)` which updates the tags in the note and the `UiState` that is handled by the fragment to update the tags (`Chip` views)
5. `TAG FIELD LOSE FOCUS`: If the field had a valid tag, call `viewModel.addTag(tag, mNote)` which updates the tags in the note and the `UiState` that is handled by the fragment to update the tags (`Chip` views). Otherwise, it doesn't do anything.

I hope this explanation does not bring more confusion. Let me know if something isn't clear or where to expand.

Reply:
❤️ Thank you so much for digging so deep to explain all that to me, much much much appreciate! Everything, seems like they are in order. Thank YOU! 🙇